kernel: assert if k_current_get is called pre-kernel

rbbrns opened the pull request:
k_current_get is not valid pre-kernel. It will return an invalid dummy thread or invalid memory. 
The invalid memory case can occur when CURRENT_THREAD_USE_TLS is enabled.

Assert in k_current_get when called pre-kernel so offending code may be identified.

k_is_pre_kernel is moved up in kernel.h to avoid needing a prototype for k_is_pre_kernel.

rbbrns committed:
kernel: assert if k_current_get is called pre-kernel

k_current_get is not valid pre-kernel. It will return an invalid dummy
thread or invalid memory. The invalid memory case can occur when
CURRENT_THREAD_USE_TLS is enabled.

Assert in k_current_get when called pre-kernel so offending code may be
identified.

k_is_pre_kernel is moved up in kernel.h to avoid needing a prototype for
k_is_pre_kernel.

Signed-off-by: Rob Barnes <robbarnes@google.com>

andyross reviewed (dismissed):
Very reasonable.  Whether exposing that symbol in kernel.h is worth it is arguable I guess; a function calls and the null check are probably about the same code size.

rbbrns commented:
Tests with USERSPACE enabled are failing because `z_sys_post_kernel` is being accessed by `k_is_pre_kernel()`. I guess `k_is_pre_kernel` could be changed to a kernel object with all user threads granted access, but now this becomes a heavier check, potentially having an impact on performance. I'm not a Zephyr kernel expert, so if anyone knows how to work around this, please comment.

rbbrns committed:
kernel: make k_is_pre_kernel safe to call from user mode

Make k_is_pre_kernel safe to call from user mode. Since
z_sys_post_kernel memory is not accessible to user threads, calling
k_is_pre_kernel would result in a memory access fault.

This change adds k_is_user_context guard. It can be assumed the system
is post-kernel if k_is_user_context is true.

Signed-off-by: Rob Barnes <robbarnes@google.com>

jhedberg commented:
@aescolar looks like a mock needed for `k_current_get()` or its new dependency, `k_is_pre_kernel()`? Note: I can't effectively work on a fix since this (again) requires a special kind of Linux environment (at least I get errors when I try to build on macOS).

jhedberg commented:
Btw, my preferred solution would be to wipe out all of the Bluetooth host unit tests (as not properly maintained and me as maintainer not even being able to build them), but I think it's too late for such a large change for Zephyr 4.3. I even had a draft PR ready doing that, but it hasn't been updated for a few months.

sjanc commented:
https://github.com/zephyrproject-rtos/zephyr/pull/98423   should fix this
